gossip/gossip.go

```diff
@@ -159,7 +159,20 @@ func New(rpcContext *rpc.Context, resolvers []resolver.Resolver) *Gossip {
 	}
 
 	// Add ourselves as a SystemConfig watcher.
-	g.is.registerCallback(KeySystemConfig, g.updateSystemConfig)
+	updateC, unregister := g.RegisterUpdateChannel(KeySystemConfig)
+	g.stopper.RunWorker(func() {
+		for {
+			select {
+			case n := <-updateC:
+				g.updateSystemConfig(n)
+				updateC <- n
```

Comment:
The need for lockstep execution definitely makes the use of callbacks more appealing than using channels that need to communicate back their completion in order to provide the ordering guarantees listed above. If we were ok with the callbacks blocking `gossip`, we could either run a goroutine that is in charge of executing callbacks that it receives in a channel through some message similar to the current `notificationRequest` (with no requestQueue), or we could just remove the goroutines from [here](https://github.com/cockroachdb/cockroach/blob/master/gossip/infostore.go#L215) and [here](https://github.com/cockroachdb/cockroach/blob/master/gossip/infostore.go#L245) and have them run on the gossip goroutine itself. If we can't permit the callbacks blocking gossip but still want to provide the 2 ordering guarantees, then I believe the only option is to buffer somewhere. I know a fixed size buffer is almost never ideal, but that might be one of the better compromises here.

I had originally attempted to follow the same pattern as before with the buffered indicator channel, but couldn't figure out a way where it could apply nicely here. The problem is that while with the `SystemConfig` indicator, the receiver knew what to query for later when it noticed the indication (`GetSystemConfig()`). However, with these key pattern registrations here, the receiver would only knows that (at least) one key matching its registered regexp pattern was updated. It would then need to re-query for all matching info each time anything was updated, instead of being alerted of the specific key or keys that had just been updated. Also, while I don't believe it matters for this situation, this approach would lose intermediate values for the same key if it was updated in quick succession between lookups.